accepted/future-releases/nnbd/feature-specification.md

```diff
@@ -495,10 +498,16 @@ of a function call produces a result of type `Never`.
 
 The static type of a `throw e` expression is `Never`.
 
-In legacy mode, an override of operator== with no explicit parameter type
+An expression of the form `e1 == e2` where the static type of `e1` and `e2` is
```

Comment:
I don't think this does what we want.  For a number of types `T`, **NonNull**(`T`) is nullable (e.g. `dynamic`, `Function`, or `FutureOr<S>` where `S` is nullable).  This formulation would not allow that invocation.  I think the `**NonNull**(`T1`) part is ok, but I think you need to use something like `T2` assignable to `S?` where `S` is the parameter type of `operator ==` on `**NonNull**(`T`).  Does that make sense?

Reply:
Of course! I changed the approach such that it uses desugaring to specify the expression in general (for the static analysis as well as the dynamic semantics), and used the `let` construct which is introduced later in the nnbd spec. The resulting expression would then rely on promotion to ensure that it is statically known that the `operator ==` is only invoked when both operands are non-null. How do you like that?